lib/rpmsg/rpmsg_virtio.c

```diff
@@ -262,96 +267,154 @@ static int _rpmsg_virtio_get_buffer_size(struct rpmsg_virtio_device *rvdev)
+						uint32_t *len, int wait)
+{
+	struct rpmsg_virtio_device *rvdev;
+	struct rpmsg_hdr *rp_hdr;
 	uint16_t idx;
 	int tick_count;
-	uint32_t buff_len;
-	int status;
-	struct metal_io_region *io;
 
 	/* Get the associated remote device for channel. */
 	rvdev = metal_container_of(rdev, struct rpmsg_virtio_device, rdev);
 
-	status = rpmsg_virtio_get_status(rvdev);
-	/* Validate device state */
-	if (!(status & VIRTIO_CONFIG_STATUS_DRIVER_OK)) {
-		return RPMSG_ERR_DEV_STATE;
-	}
-
 	if (wait)
 		tick_count = RPMSG_TICK_COUNT / RPMSG_TICKS_PER_INTERVAL;
 	else
 		tick_count = 0;
 
 	while (1) {
-		int avail_size;
-
 		/* Lock the device to enable exclusive access to virtqueues */
 		metal_mutex_acquire(&rdev->lock);
-		avail_size = _rpmsg_virtio_get_buffer_size(rvdev);
-		if (size <= avail_size)
-			buffer = rpmsg_virtio_get_tx_buffer(rvdev, &buff_len,
-							    &idx);
+		rp_hdr = rpmsg_virtio_get_tx_buffer(rvdev, len, &idx);
 		metal_mutex_release(&rdev->lock);
-		if (buffer || !tick_count)
+		if (rp_hdr || !tick_count)
 			break;
-		if (avail_size != 0)
-			return RPMSG_ERR_BUFF_SIZE;
 		metal_sleep_usec(RPMSG_TICKS_PER_INTERVAL);
 		tick_count--;
 	}
-	if (!buffer)
-		return RPMSG_ERR_NO_BUFF;
+
+	if (!rp_hdr)
+		return NULL;
+
+	/* Store the index into the reserved field to be used when sending */
+	rp_hdr->reserved = idx;
+
+	/* Actual data buffer size is vring buffer size minus rpmsg header length */
+	*len -= sizeof(struct rpmsg_hdr);
+	return RPMSG_LOCATE_DATA(rp_hdr);
+}
+
+static int rpmsg_virtio_send_offchannel_nocopy(struct rpmsg_device *rdev,
+					       uint32_t src, uint32_t dst,
+					       const void *data, int len)
+{
+	struct rpmsg_virtio_device *rvdev;
+	struct rpmsg_hdr rp_hdr;
+	struct rpmsg_hdr *hdr;
+	uint16_t idx;
+	int status;
+	struct metal_io_region *io;
+
+	/* Get the associated remote device for channel. */
+	rvdev = metal_container_of(rdev, struct rpmsg_virtio_device, rdev);
+
+	hdr = RPMSG_LOCATE_HDR(data);
+	/* The reserved field contains buffer index */
+	idx = hdr->reserved;
 
 	/* Initialize RPMSG header. */
 	rp_hdr.dst = dst;
 	rp_hdr.src = src;
-	rp_hdr.len = size;
+	rp_hdr.len = len;
 	rp_hdr.reserved = 0;
+	rp_hdr.flags = 0;
 
 	/* Copy data to rpmsg buffer. */
 	io = rvdev->shbuf_io;
-	status = metal_io_block_write(io, metal_io_virt_to_offset(io, buffer),
+	status = metal_io_block_write(io, metal_io_virt_to_offset(io, hdr),
 				      &rp_hdr, sizeof(rp_hdr));
 	RPMSG_ASSERT(status == sizeof(rp_hdr), "failed to write header\r\n");
 
-	status = metal_io_block_write(io,
-				      metal_io_virt_to_offset(io,
-				      RPMSG_LOCATE_DATA(buffer)),
-				      data, size);
-	RPMSG_ASSERT(status == size, "failed to write buffer\r\n");
 	metal_mutex_acquire(&rdev->lock);
 
 	/* Enqueue buffer on virtqueue. */
-	status = rpmsg_virtio_enqueue_buffer(rvdev, buffer, buff_len, idx);
+	status = rpmsg_virtio_enqueue_buffer(rvdev, hdr, 0, idx);
```

Comment:
should you use len instead of "0"? you can set len default value if len is not set here.

Reply:
Your mean 'len' parameter pass from caller? We can't use it here since virito layer need the whole buffer length, not the real playload length.

Reply:
> Your mean 'len' parameter pass from caller? We can't use it here since virito layer need the whole buffer length, not the real playload length.

As both zero copy and non-zero copy calling this function, for non-zero copy case, it can get the buffer length.

Reply:
Ok, the new patch remove the dependence between zero copy and non-zero copy. Please take a look.